SU2_CFD/src/numerics_direct_turbulent.cpp

```diff
@@ -1222,42 +1222,320 @@ void CSourcePieceWise_TurbSST::ComputeResidual(su2double *val_residual, su2doubl
+
+   /* if using UQ methodolgy, calculate production using perturbed Reynolds stress matrix */
+
+   if (config->GetUsing_UQ()){
+    pw = PerturbedStrainMag * PerturbedStrainMag - 2.0/3.0*zeta*diverg;
+   }
+   else {
+     pw = StrainMag_i*StrainMag_i - 2.0/3.0*zeta*diverg;
+   }
+   pw = max(pw,0.0);
+
+   val_residual[0] += pk*Volume;
+   val_residual[1] += alfa_blended*Density_i*pw*Volume;
+
+   /*--- Dissipation ---*/
+
+   val_residual[0] -= beta_star*Density_i*TurbVar_i[1]*TurbVar_i[0]*Volume;
+   val_residual[1] -= beta_blended*Density_i*TurbVar_i[1]*TurbVar_i[1]*Volume;
+
+   /*--- Cross diffusion ---*/
+
+   val_residual[1] += (1.0 - F1_i)*CDkw_i*Volume;
+
+   /*--- Implicit part ---*/
+
+   val_Jacobian_i[0][0] = -beta_star*TurbVar_i[1]*Volume;
+   val_Jacobian_i[0][1] = -beta_star*TurbVar_i[0]*Volume;
+   val_Jacobian_i[1][0] = 0.0;
+   val_Jacobian_i[1][1] = -2.0*beta_blended*TurbVar_i[1]*Volume;
   }
   
   AD::SetPreaccOut(val_residual, nVar);
   AD::EndPreacc();
 
 }
+
+void CSourcePieceWise_TurbSST::GetMeanRateOfStrainMatrix(su2double **S_ij)
+{
+    /* --- Calculate the rate of strain tensor, using mean velocity gradients --- */
+
+  if (nDim == 3){
+    S_ij[0][0] = PrimVar_Grad_i[1][0];
+    S_ij[1][1] = PrimVar_Grad_i[2][1];
+    S_ij[2][2] = PrimVar_Grad_i[3][2];
+    S_ij[0][1] = 0.5 * (PrimVar_Grad_i[1][1] + PrimVar_Grad_i[2][0]);
+    S_ij[0][2] = 0.5 * (PrimVar_Grad_i[1][2] + PrimVar_Grad_i[3][0]);
+    S_ij[1][2] = 0.5 * (PrimVar_Grad_i[2][2] + PrimVar_Grad_i[3][1]);
+    S_ij[1][0] = S_ij[0][1];
+    S_ij[2][1] = S_ij[1][2];
+    S_ij[2][0] = S_ij[0][2];
+  }
+  else {
+    S_ij[0][0] = PrimVar_Grad_i[1][0];
+    S_ij[1][1] = PrimVar_Grad_i[2][1];
+    S_ij[2][2] = 0.0;
+    S_ij[0][1] = 0.5 * (PrimVar_Grad_i[1][1] + PrimVar_Grad_i[2][0]);
+    S_ij[0][2] = 0.0;
+    S_ij[1][2] = 0.0;
+    S_ij[1][0] = S_ij[0][1];
+    S_ij[2][1] = S_ij[1][2];
+    S_ij[2][0] = S_ij[0][2];
+
+  }
+}
+
+void CSourcePieceWise_TurbSST::SetReynoldsStressMatrix(su2double turb_ke){
+  unsigned short iDim, jDim;
+  su2double **S_ij = new su2double* [3];
+  su2double divVel = 0;
+  su2double TWO3 = 2.0/3.0;
+
+
+
+  for (iDim = 0; iDim < 3; iDim++){
+    S_ij[iDim] = new su2double [3];
+  }
+
+  GetMeanRateOfStrainMatrix(S_ij);
+
+    /* --- Using rate of strain matrix, calculate Reynolds stress tensor --- */
+
+  for (iDim = 0; iDim < 3; iDim++){
+    divVel += S_ij[iDim][iDim];
+  }
+
+  for (iDim = 0; iDim < 3; iDim++){
+    for (jDim = 0; jDim < 3; jDim++){
+      MeanReynoldsStress[iDim][jDim] = TWO3 * turb_ke * delta3[iDim][jDim]
+      - Eddy_Viscosity_i / Density_i * (2 * S_ij[iDim][jDim] - TWO3 * divVel * delta3[iDim][jDim]);
+    }
+  }
+
+  for (iDim = 0; iDim < 3; iDim++)
+    delete [] S_ij[iDim];
+  delete [] S_ij;
+}
+
+void CSourcePieceWise_TurbSST::SetPerturbedRSM (su2double turb_ke, CConfig *config){
+
+  unsigned short iDim,jDim;
+  su2double **A_ij;
+  su2double **Eig_Vec;
```

Comment:
Same comment about mem allocations w.r.t. being class data and hard-coded '3' values.

Reply:
Addressed in other comments